### ISSUE TITLE
Some elements in the original IFC model cannot be displayed after uploading to Bimserver

### ISSUE THREAD

**Issue by Billy8424** [opensourceBIM/BIMserver#1334]
Hello, we tried to upload a IFC model to BIMserver, but some equipment elements cannot be displayed after uploading to BIMserver, these elements can be displayed in the original IFC model. Please check the screenshots below, could you please have a check and let us know the reason? Thanks a lot. 

Below is the original IFC model, we can see the equipment in the hall. 
![Original IFC model](https://github.com/user-attachments/assets/054ce116-39b5-43ad-bc92-1933eff9b4d2)

This is the model after uploading to BIMserver, it's empty in the hall, the equipment is missing. 
![IFC model in Bimserver](https://github.com/user-attachments/assets/acdfe395-1f0f-43f1-b99d-4378131ae2b7)

**Comment by hlg** [comment]
It is not possible to tell where the elements are lost from those two screenshots. The IFC data undergoes a couple of steps until the final display: upload to BIMserver and deserialization, storage in the database, potentially creation of triangulated geometry, storage of the geometry, download from BIMserver, potentially filter by a query, serialization to IFC or a viewer-specific format, parsing in the viewer, potentially mapping to a scene graph or other viewer model, rendering. Please narrow down where the issue appears so that we can see whether this is actually a BIMserver issue and hopefully help solving it. Here are some questions to aid you:

- Which BIMserver version are you using? How do you check-in? Via BIMvie.ws GUI? With which deserializer? If applicable: which API method?
- Is the data complete after checking it into the BIMserver? Can you find the elements that are missing later?
- Which viewer are you using? I can't recognize it from the screenshot.
- Which data is the viewer taking from the BIMserver? The original IFC only or also the triangulated geometry? And in which format? Which serializer? With a query?
- If relevant: Which render engine and implementation version are you using? Is the geometry complete after triangulation or do elements fail to triangulate?
- Is the data complete after download and before display in the viewer?

**Comment by Billy8424** [comment]
Hello,

Thanks for your feedback.

Maybe the most efficient way is I send the IFC model we tested to you, and you can also have a try? You can download the IFC model through this URL: https://we.tl/t-wdBlSFBAZA <https://we.tl/t-wdBlSFBAZA?utm_campaign=TRN_TDL_05&utm_source=sendgrid&utm_medium=email&trk=TRN_TDL_05>

In the original IFC model, we can see the equipment inside the 3 halls, like the elements that marked in Green in the screenshot below. But after uploading to the BIM Server, these elements cannot be displayed. Please help to check the issue and let us know the reason.

Thank you very much.

***@***.***

Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: hlg ***@***.***>
发送时间: 2024年10月30日 0:11
收件人: opensourceBIM/BIMserver ***@***.***>
抄送: Billy Cai ***@***.***>; Author ***@***.***>
主题: Re: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)


It is not possible to tell where the elements are lost from those two screenshots. The IFC data undergoes a couple of steps until the final display: upload to BIMserver and deserialization, storage in the database, potentially creation of triangulated geometry, storage of the geometry, download from BIMserver, potentially filter by a query, serialization to IFC or a viewer-specific format, parsing in the viewer, potentially mapping to a scene graph or other viewer model, rendering. Please narrow down where the issue appears so that we can see whether this is actually a BIMserver issue and hopefully help solving it. Here are some questions to aid you:

  *   Which BIMserver version are you using? How do you check-in? Via BIMvie.ws GUI? With which deserializer? If applicable: which API method?
  *   Is the data complete after checking it into the BIMserver? Can you find the elements that are missing later?
  *   Which viewer are you using? I can't recognize it from the screenshot.
  *   Which data is the viewer taking from the BIMserver? The original IFC only or also the triangulated geometry? And in which format? Which serializer? With a query?
  *   If relevant: Which render engine and implementation version are you using? Is the geometry complete after triangulation or do elements fail to triangulate?
  *   Is the data complete after download and before display in the viewer?

**Comment by hlg** [comment]
Thank you for providing the IFC data. With a  current BIMserver (1.5.186-SNAPSHOT) and IfcOpenShell (v0.7.0-c7830e9) version, and using BIMvie.ws as a viewer, I cannot reproduce the issue. See below, where you can see the machinery with and without its housing.

![housing-on-s](https://github.com/user-attachments/assets/493bc5f4-a688-4470-8f75-b2fdb8f26f18) ![housing-off-s](https://github.com/user-attachments/assets/a04f40d1-7841-4deb-9ab5-3df07eedc22a)

**Comment by Billy8424** [comment]
Hello，

Thanks for your feedback.

Ok, we are now using the older version with ifcopenshellplugin-0.5.91 and Bim Server 1.5.182. Could you please let us know how to download and deploy the new version as yours BIMserver (1.5.186-SNAPSHOT) and IfcOpenShell (v0.7.0-c7830e9)?

Thank you.

Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: hlg ***@***.***>
发送时间: 2024年11月5日 18:40
收件人: opensourceBIM/BIMserver ***@***.***>
抄送: Billy Cai ***@***.***>; Author ***@***.***>
主题: Re: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)


Thank you for providing the IFC data. With a current BIMserver (1.5.186-SNAPSHOT) and IfcOpenShell (v0.7.0-c7830e9) version, and using BIMvie.ws as a viewer, I cannot reproduce the issue. See below, where you can see the machinery with and without its housing.

housing-on-s.png (view on web)<https://github.com/user-attachments/assets/493bc5f4-a688-4470-8f75-b2fdb8f26f18> housing-off-s.png (view on web)<https://github.com/user-attachments/assets/a04f40d1-7841-4deb-9ab5-3df07eedc22a>

**Comment by hlg** [comment]
Are you aware that your email replies are visible in the Github issue tracker and clutter the discussion? Please go to the Github web interface and cleanup your comments.

Snapshot versions cannot be downloaded. You can build it yourself or wait for the next release. For now, you can also try the latest released BIMserver version (1.5.185) with the last IfcOpenShell v0.6.0 version: 721fe47.

**Comment by Billy8424** [comment]
Hello,

Thanks for your reminder, I have deleted some comments in Github.

Could you please let me know the planned date for the next release of 1.5.186?

We have already tried the latest released version 1.5.185, but also cannot see these elements after uploading to the BIM server, could you please help to also have a try with 1.5.185? We see the last commit in Github's sources was yesterday for 185-Snapshot, is there any difference between 185-Snapshot and 1.5.185?

Thank you.

Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: hlg ***@***.***>
发送时间: 2024年11月13日 0:53
收件人: opensourceBIM/BIMserver ***@***.***>
抄送: Billy Cai ***@***.***>; Author ***@***.***>
主题: Re: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)


Are you aware that your email replies are visible in the Github issue tracker and clutter the discussion? Please go to the Github web interface and cleanup your comments.

Snapshot versions cannot be downloaded. You can build it yourself or wait for the next release. For now, you can also try the latest released BIMserver version (1.5.185) with the last IfcOpenShell v0.6.0 version: 721fe47.

**Comment by Billy8424** [comment]
Hello，

Thanks for your feedback.

Seems the default IfcOpenShell with BIM Server 185 is V. 0.5.91, we tried to compile IfcOpenShell V.0.6.0 but failed because lack of many dependencies. Could you please suggest how shall we use the IfcOpenShell V.0.6.0 721fe47 for BIM Server 185?

And do you have any planned date for release the BIM Server 186?

Thank you.
Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: zqf ***@***.***>
发送时间: 2024年11月15日 16:40
收件人: opensourceBIM/BIMserver ***@***.***>
抄送: Billy Cai ***@***.***>; Author ***@***.***>
主题: Re: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)


have you also tried to use the last IfcOpenShell v0.6.0 version: 721fe47 as hlg mentioned? because the default version of the IfcOpenShell is 2f3c79a

**Comment by zaqifathis** [comment]
You are correct; the default IfcOpenShell for BIMserver 1.5.185 is version 0.5.91.

I successfully built the [IfcOpenShell plugin](https://github.com/opensourceBIM/IfcOpenShell-BIMserver-plugin). It worked well for me. You can build from this commit  `94ece7c`  and change the version to `721fe47`[here](https://github.com/opensourceBIM/IfcOpenShell-BIMserver-plugin/commit/7157e503868fe52d51c6aa543c87d9b766c53235#diff-c50dc7f7fd4c00a87162cddecd539384da3ef7cec7a7713396466fd0b993bdb6)

**Comment by Billy8424** [comment]
Hello,

Following up my last email, the screenshot below shows the version of all the plugins that in our current BIM Server. Just FYI.

Thank you.

***@***.***

Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: Billy Cai
发送时间: 2024年11月25日 16:35
收件人: opensourceBIM/BIMserver ***@***.***>; opensourceBIM/BIMserver ***@***.***>
抄送: Author ***@***.***>
主题: 回复: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)

Hello,

Thanks for your feedback.

Yes, we tried to use BIM Server 185 and IfcOpenShell plugin 0.5.94 to upload this IFC model, but still cannot see the equipment elements inside after uploading to the BIM Server.

As you mentioned in the previous email, ‘Just change the commit SHA to 721fe47 in the plugin settings.’ Sorry, could you please tell me where to change the plugin setting?  I cannot find where to see and change this information.

So you meant that you also tried to upload this IFC model to BIM Server 185 and IfcOpenShell plugin 0.5.94, and the equipment can be seen?

Thank you.


Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: hlg ***@***.******@***.***>>
发送时间: 2024年11月18日 21:49
收件人: opensourceBIM/BIMserver ***@***.******@***.***>>
抄送: Billy Cai ***@***.******@***.***>>; Author ***@***.******@***.***>>
主题: Re: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)


For BIMserver 1.5.185, the last compatible IfcOpenShell plugin version is 0.5.94 which uses the IfcOpenShell build with commit SHA 2f3c79a (from the v0.6.0 branch) by default. You don't have to build neither IfcOpenShell itself nor the IfcOpenShell plugin. Just change the commit SHA to 721fe47 in the plugin settings. This has been explained multiple times in previous issues. However, even that is not necessary. As far as I can see, IfcOpenShell build 2f3c79a works just as fine for your sample. Have you even tried BIMserver 1.5.185 with the default plugins at all?

Note that we don't provide individual email support, but we are having a public discussion around a potential issue in Github. Please keep the discussion readable. Don't delete your comments in the Github web interface, but remove the email headers, footers and quotes or directly participate in the discussion via Github web interface instead of email reply. If you want to send confidential test data or disclose any other information useful for BIMserver development, but not meant for the general public, then you can email to ***@***.******@***.***>.

**Comment by zaqifathis** [comment]
Please remove the email header, etc.. from Github web interface as it hard to follow the discussion.

As explained before, to change the commit SHA, you just need to go to Server Setting tab > Installed Plugin Bundle > select IfcOpenShell-BIMserver-plugin > select IfcOpenShell.

![Screenshot 2024-11-27 111934](https://github.com/user-attachments/assets/5c1d1fca-9d73-4d75-a443-d0827b5735e4)

**Comment by hlg** [comment]
I repeat: For both BIMserver 1.5.185 and 1.5.186 and the compatible IfcOpenShell plugins with their default IfcOpenShell commit sha settings, I cannot reproduce your issue - I can see the engine geometries in BIMvie.ws with the included older BIMsurfer version. There should be no need to change the commit sha or anything else in the setup.

If the issue persists for you, but we can not reproduce it, we need more information to find the cause. Please look at my first comment to this issue with a list of questions. Most importantly: Are you trying to see the geometry in BIMvie.ws or with a different viewer? Your screenshot in the original post does not look like BIMvie.ws, but something else with the grey background and the navigation controls.

**Comment by Billy8424** [comment]
Hello,

Thanks for your feedback.

We have tried to upgrade to BIM Server 186 and IfcOpenShell v0.7.0-c7830e9, should be same to your testing environment. Please see the screenshots below.

***@***.***

***@***.***

***@***.***

But after we uploaded the IFC model, we still cannot see the equipment inside, we just used the BIMvie.ws to view the BIM model. Please see the screenshot below.  Since you are able to see the equipment with the same version of BIM Server and IfcOpenShell. Do you have any other suggestion why we cannot see it?

Thank you very much.

***@***.***


Best Regards
Billy Cai 蔡文俊
Operations Director
Siveco China | Bluebee Technologies
Room 20J, Zao Fong Universe Building
No.1800 Zhong Shan Road (W)
Shanghai 200235
P.R. China
Follow us on ***@***.*** <https://www.sivecochina.com/sites/all/themes/siveco/images/QRcode.png>  [linkedin - 2] <linkedin.com/company/siveco-china>
T +86 21 6440 3226 | M +86 13564566073
E ***@***.******@***.***>
www.sivecochina.com<http://www.sivecochina.com/> | newsletter.sivecochina.com

***@***.***




发件人: hlg ***@***.***>
发送时间: 2024年11月28日 4:28
收件人: opensourceBIM/BIMserver ***@***.***>
抄送: Billy Cai ***@***.***>; Author ***@***.***>
主题: Re: [opensourceBIM/BIMserver] Some elements in the original IFC model cannot be displayed after uploading to Bimserver (Issue #1334)


I repeat: For both BIMserver 1.5.185 and 1.5.186 and the compatible IfcOpenShell plugins with their default IfcOpenShell commit sha settings, I cannot reproduce your issue - I can see the engine geometries in BIMvie.ws with the included older BIMsurfer version. There should be no need to change the commit sha or anything else in the setup.

If the issue persists for you, but we can not reproduce it, we need more information to find the cause. Please look at my first comment to this issue with a list of questions. Most importantly: Are you trying to see the geometry in BIMvie.ws or with a different viewer? Your screenshot in the original post does not look like BIMvie.ws, but something else with the grey background and the navigation controls.

**Comment by hlg** [comment]
Sorry, I cannot see any screenshots. For further investigation, you can follow the steps of the data from upload to the final visualization as outlined in the first reply already: 1. verify that the missing elements are there in BIMserver, for example search for a GUID using the "Browse" tab in BIMvie.ws, 2. check the geometry generation report for errors ("Extended data" tab), 3. make sure your viewer is capable, functioning with errors and not ignoring or hiding any elements.

**Comment by Billy8424** [comment]
hello, thanks for your feedback. 
From another BIM viewer, I got the GUID of this equipment, see below screenshot
<img width="960" alt="another viewer" src="https://github.com/user-attachments/assets/4a78a454-bab8-4e63-8e72-004d8192af3a">
And in BIMvie.ws, I followed your instruction, search this GUID in the Browse tab, can find this element. 
<img width="621" alt="GUID in BIMvie" src="https://github.com/user-attachments/assets/dc8d0846-20c7-4936-b7f0-baa8fec4eecd">
But in the BIMvie.ws viewer, I still cannot see the equipment, even in the structure seems I can also find this element, please see the screenshot below
<img width="684" alt="No equipment in BIMvie" src="https://github.com/user-attachments/assets/4e47d03b-4bf7-4d4c-b1b3-acb858faeed6">
We have already upgraded the BIM server to 186 and the IfcOpenShell to v0.7.0-c7830e9. As you suggested in the last email, could you please let me know how to 'make sure your viewer is capable, functioning with errors and not ignoring or hiding any elements'?
By the way, is it possible for you to have a remote check to our environment or we can have a MS Teams meeting to review it together? 
Thank you very much for your support.

**Comment by hlg** [comment]
The first step shows us that this is not a core BIMserver error or basic issue with the IFC file. The second step - check geometry generation report - is the next important now, because it will tell us whether the geometry for these elements was processed correctly. The third step I mentioned pertained to you possibly using a different viewer, for which I could not give any advise. With the BIMsurfer in BIMvie.ws this step is obsolete.

**Comment by hlg** [comment]
We were able to reproduce this issue with the BIMserver war in Tomcat. Although there is no exception reported in the log or the geometry report, there seems to be an irregularity in the geometry generation report, where one job with over 60 entities of type IfcBuildingElementProxy is showing 0 triangles been created using a negative amount of time. This can only be explained if the geometry engine terminated in a way such that the end time could not be recorded in BIMserver. This same job also takes a very long time - almost 10 minutes - in  my setup with the BIMserver jar where it eventually succeeds.

**Comment by hlg** [comment]
Finally I could reproduce this with the BIMserver JAR as well, not just in Tomcat, and resolve by increasing the heap size. This particular file produces one very large geometry generation job (16 elements of type IfcBuildingElementProxy amounting to an exceptional IFC-SPF of around 200 MB passed to the render engine). This job includes those entities which are ultimately missing from the results. With smaller heap sizes, which are fine otherwise, this triggers an OutOfMemoryError when fetching the IFC from the database or later, when processing the results. In the geometry runner thread, this error slipped through the exception handling and thus yielded the strange result with no visible errors in the geometry generation report or logs, and just the negative duration instead. The logging has been improved, so as to easier find a remedy if this occurs again in the future.

@Billy8424, please check if increaing heap fixes this and close this issue if you can confirm.

**Comment by hlg** [comment]
Confirmed to be fixed with more heap.
